File: salesflow-app/src/backend/app/domain/goals/vertical_adapter.py
"""
╔════════════════════════════════════════════════════════════════════════════╗
║  SALES FLOW AI - BASE VERTICAL ADAPTER                                    ║
║  Abstrakte Basisklasse für Branchen-Adapter                               ║
╚════════════════════════════════════════════════════════════════════════════╝

Jeder Vertical-Adapter implementiert diese Schnittstelle:
  → Network Marketing: Rang → Volumen → Kunden/Partner
  → Immobilien: Provision → Deals → Leads
  → Coaching: MRR → Klienten → Discovery Calls
  → etc.

Sync mit TypeScript:
  → src/services/verticalAdapters/baseAdapter.ts
"""

from abc import ABC, abstractmethod
from typing import List

from .types import (
    GoalInput,
    GoalBreakdown,
    DailyFlowConfig,
    KpiDefinition,
)


class BaseVerticalAdapter(ABC):
    """
    Abstrakte Basisklasse für Vertical-Adapter.
    
    Jeder Adapter muss folgende Methoden implementieren:
    - vertical_id: Eindeutige ID der Branche
    - get_label(): Anzeigename
    - compute_goal_breakdown(): Ziel → Breakdown
    - get_default_conversion_config(): Standard-Conversion-Rates
    - get_kpi_definitions(): KPIs für Dashboard
    """
    
    @property
    @abstractmethod
    def vertical_id(self) -> str:
        """Eindeutige ID der Branche (z.B. 'network_marketing')."""
        pass
    
    @abstractmethod
    def get_label(self) -> str:
        """Anzeigename der Branche (z.B. 'Network Marketing / MLM')."""
        pass
    
    @abstractmethod
    def compute_goal_breakdown(self, goal_input: GoalInput) -> GoalBreakdown:
        """
        Berechnet aus einem Ziel den vollständigen Breakdown.
        
        Args:
            goal_input: Ziel-Definition (Art, Wert, Zeitraum, etc.)
            
        Returns:
            GoalBreakdown mit allen berechneten Werten
        """
        pass
    
    @abstractmethod
    def get_default_conversion_config(self) -> DailyFlowConfig:
        """
        Standard-Conversion-Rates für diese Branche.
        
        Beispiel Network Marketing:
            - 20% Kontakte → Kunde
            - 5% Kontakte → Partner
            - 3 Follow-ups pro Kunde
            - 5 Follow-ups pro Partner
        """
        pass
    
    @abstractmethod
    def get_kpi_definitions(self) -> List[KpiDefinition]:
        """
        KPI-Definitionen für das Dashboard.
        
        Gibt eine Liste von KPIs zurück, die für diese Branche relevant sind.
        """
        pass
    
    # ═══════════════════════════════════════════════════════════════════════
    # HELPER METHODS (optional zu überschreiben)
    # ═══════════════════════════════════════════════════════════════════════
    
    def compute_daily_targets(
        self,
        breakdown: GoalBreakdown,
        config: DailyFlowConfig,
    ) -> dict:
        """
        Berechnet tägliche Aktivitäts-Targets aus dem Breakdown.
        
        Kann von Subklassen überschrieben werden für spezielle Logik.
        """
        weeks = breakdown.timeframe_months * 4.33
        working_days = weeks * config.working_days_per_week
        
        # Kontakte basierend auf Conversion Rate
        contacts_needed = breakdown.primary_units / config.contact_to_primary_unit
        contacts_per_day = contacts_needed / working_days
        
        # Follow-ups
        total_followups = (
            breakdown.primary_units * config.followups_per_primary +
            breakdown.secondary_units * config.followups_per_secondary
        )
        followups_per_day = total_followups / working_days
        
        # Reaktivierungen
        reactivations_per_day = contacts_per_day * config.reactivation_share
        
        return {
            "daily": {
                "new_contacts": round(contacts_per_day),
                "followups": round(followups_per_day),
                "reactivations": round(reactivations_per_day),
            },
            "weekly": {
                "new_contacts": round(contacts_per_day * config.working_days_per_week),
                "followups": round(followups_per_day * config.working_days_per_week),
                "reactivations": round(reactivations_per_day * config.working_days_per_week),
                "primary_units": round(breakdown.primary_units / weeks, 1),
                "secondary_units": round(breakdown.secondary_units / weeks, 1),
            },
        }
    
    def format_breakdown_summary(self, breakdown: GoalBreakdown) -> str:
        """Formatiert den Breakdown als lesbaren Text."""
        return f"""
🎯 Ziel-Breakdown ({self.get_label()})

📊 Benötigtes Volumen: {breakdown.required_volume:,.0f}
   • Pro Monat: {breakdown.per_month_volume:,.0f}
   • Pro Woche: {breakdown.per_week_volume:,.0f}
   • Pro Tag: {breakdown.per_day_volume:,.0f}

👥 Geschätzte Einheiten:
   • Primär (Kunden): {breakdown.primary_units:,.0f}
   • Sekundär (Partner): {breakdown.secondary_units:,.0f}

📝 {breakdown.notes}
        """.strip()

File: salesflow-app/src/types/verticalAdapter.ts
/**
 * ╔════════════════════════════════════════════════════════════════════════════╗
 * ║  SALES FLOW AI - VERTICAL ADAPTER TYPES                                   ║
 * ║  Type-Definitionen für Goal-Berechnung (sync mit Python-Backend)          ║
 * ╚════════════════════════════════════════════════════════════════════════════╝
 * 
 * Sync mit Python:
 *   → backend/app/domain/goals/types.py
 */

import { z } from 'zod';

// ═══════════════════════════════════════════════════════════════════════════
// ENUMS
// ═══════════════════════════════════════════════════════════════════════════

export const GoalKindSchema = z.enum([
  'income',     // Ziel-Einkommen (€/Monat)
  'rank',       // Rang erreichen
  'volume',     // Volumen-Ziel (Credits/PV)
  'customers',  // Anzahl Kunden
  'partners',   // Anzahl Partner
  'deals',      // Abschlüsse (Immobilien, Solar)
]);
export type GoalKind = z.infer<typeof GoalKindSchema>;

// Re-export VerticalId from existing types
export { VerticalIdSchema, type VerticalId } from '../config/verticals/types';

// ═══════════════════════════════════════════════════════════════════════════
// INPUT TYPES
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Eingabe für Goal-Berechnung.
 * 
 * @example
 * const input: GoalInput = {
 *   vertical_id: 'network_marketing',
 *   goal_kind: 'income',
 *   target_value: 2000,  // 2.000€/Monat
 *   timeframe_months: 6,
 *   vertical_meta: { comp_plan_id: 'zinzino' }
 * };
 */
export const GoalInputSchema = z.object({
  vertical_id: z.string(),
  goal_kind: GoalKindSchema,
  target_value: z.number(),
  timeframe_months: z.number().min(1).max(60),
  current_value: z.number().default(0),
  vertical_meta: z.record(z.any()).default({}),
});
export type GoalInput = z.infer<typeof GoalInputSchema>;

/**
 * Konfiguration für Daily Flow Berechnungen.
 */
export const DailyFlowConversionConfigSchema = z.object({
  working_days_per_week: z.number().default(5),
  contact_to_primary_unit: z.number().default(0.20),    // 20% Kontakte → Kunden
  contact_to_secondary_unit: z.number().default(0.05),  // 5% Kontakte → Partner
  followups_per_primary: z.number().default(3),
  followups_per_secondary: z.number().default(5),
  reactivation_share: z.number().default(0.20),
});
export type DailyFlowConversionConfig = z.infer<typeof DailyFlowConversionConfigSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// OUTPUT TYPES
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Ergebnis der Goal-Berechnung.
 */
export const GoalBreakdownSchema = z.object({
  vertical_id: z.string(),
  goal_kind: GoalKindSchema,
  timeframe_months: z.number(),
  
  // Primäre & sekundäre Einheiten
  primary_units: z.number(),
  secondary_units: z.number(),
  
  // Volumen-Aufschlüsselung
  required_volume: z.number(),
  per_month_volume: z.number(),
  per_week_volume: z.number(),
  per_day_volume: z.number(),
  
  // Vertical-spezifische Details
  vertical_details: z.record(z.any()).default({}),
  
  // Notizen/Erklärung
  notes: z.string().default(''),
});
export type GoalBreakdown = z.infer<typeof GoalBreakdownSchema>;

/**
 * Tägliche Aktivitäts-Ziele.
 */
export interface DailyFlowTargets {
  new_contacts: number;
  followups: number;
  reactivations: number;
}

/**
 * Wöchentliche Aktivitäts-Ziele.
 */
export interface WeeklyFlowTargets {
  new_contacts: number;
  followups: number;
  reactivations: number;
  primary_units: number;
  secondary_units: number;
}

/**
 * Kombinierte Daily/Weekly Targets.
 */
export interface FlowTargets {
  daily: DailyFlowTargets;
  weekly: WeeklyFlowTargets;
}

// ═══════════════════════════════════════════════════════════════════════════
// KPI DEFINITION
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Definition eines KPIs für das Dashboard.
 */
export const AdapterKpiDefinitionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  unit: z.enum(['per_day', 'per_week', 'per_month', 'total']),
  icon: z.string(),   // Lucide Icon Name
  color: z.string(),  // Tailwind Color Name
});
export type AdapterKpiDefinition = z.infer<typeof AdapterKpiDefinitionSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// ADAPTER INTERFACE
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Interface für Vertical-Adapter.
 * 
 * Jeder Adapter implementiert diese Schnittstelle:
 * - Network Marketing: Rang → Volumen → Kunden/Partner
 * - Immobilien: Provision → Deals → Leads
 * - Coaching: MRR → Klienten → Discovery Calls
 */
export interface IVerticalAdapter {
  /** Eindeutige ID der Branche */
  readonly verticalId: string;
  
  /** Anzeigename der Branche */
  getLabel(): string;
  
  /** Berechnet Goal Breakdown */
  computeGoalBreakdown(goalInput: GoalInput): GoalBreakdown;
  
  /** Standard-Conversion-Rates */
  getDefaultConversionConfig(): DailyFlowConversionConfig;
  
  /** KPIs für Dashboard */
  getKpiDefinitions(): AdapterKpiDefinition[];
  
  /** Berechnet Daily/Weekly Targets aus Breakdown */
  computeFlowTargets(breakdown: GoalBreakdown, config?: DailyFlowConversionConfig): FlowTargets;
  
  /** Formatiert Breakdown als Text */
  formatBreakdownSummary(breakdown: GoalBreakdown): string;
}

// ═══════════════════════════════════════════════════════════════════════════
// DEFAULT CONFIG
// ═══════════════════════════════════════════════════════════════════════════

export const DEFAULT_CONVERSION_CONFIG: DailyFlowConversionConfig = {
  working_days_per_week: 5,
  contact_to_primary_unit: 0.20,
  contact_to_secondary_unit: 0.05,
  followups_per_primary: 3,
  followups_per_secondary: 5,
  reactivation_share: 0.20,
};

